mos_tests/environment/ssh.py

```diff
@@ -0,0 +1,270 @@
+import logging
+import os
+import paramiko
+import posixpath
+import stat
+
+
+logger = logging.getLogger(__name__)
+
+
+class SFTPNotApplicableError(ValueError):
+    def __str__(self):
+        message = "SFTP is not applicable for SSHClient."
+        return message
+
+
+class CalledProcessError(Exception):
+    def __init__(self, command, returncode, output=None):
+        self.returncode = returncode
+        self.cmd = command
+        self.output = output
+
+    def __str__(self):
+        message = "Command '%s' returned non-zero exit status %s" % (
+            self.cmd, self.returncode)
+        if self.output:
+            message += "\n%s" % '\n'.join(self.output)
+        return message
+
+
+class SSHClient(object):
+    class get_sudo(object):
+        def __init__(self, ssh):
+            self.ssh = ssh
+
+        def __enter__(self):
+            self.ssh.sudo_mode = True
+
+        def __exit__(self, exc_type, value, traceback):
+            self.ssh.sudo_mode = False
+
+    def __init__(self, host, port=22, username=None, password=None,
+                 private_keys=None, use_sftp=True):
+        self.host = str(host)
+        self.port = int(port)
+        self.username = username
+        self.password = password
+        if not private_keys:
+            private_keys = []
+        self.private_keys = private_keys
+
+        self.sudo_mode = False
+        self.sudo = self.get_sudo(self)
+        self.use_sftp = use_sftp
+
+        self.reconnect()
+
+    def clear(self):
+        if self.use_sftp:
+            try:
+                self._sftp.close()
+            except Exception:
+                logger.exception("Could not close sftp connection")
+
+        try:
+            self._ssh.close()
+        except Exception:
+            logger.exception("Could not close ssh connection")
+
+    def __del__(self):
+        self.clear()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *err):
+        self.clear()
+
+    def connect(self):
+        logging.debug(
+            "Connect to '%s:%s' as '%s:%s'" % (
+                self.host, self.port, self.username, self.password))
+        for private_key in self.private_keys:
+            try:
+                return self._ssh.connect(
+                    self.host, port=self.port, username=self.username,
+                    password=self.password, pkey=private_key)
+            except paramiko.AuthenticationException:
+                continue
+        if self.private_keys:
+            logging.error("Authentication with keys failed")
+
+        return self._ssh.connect(
+            self.host, port=self.port, username=self.username,
+            password=self.password)
+
+    def reconnect(self):
+        self._ssh = paramiko.SSHClient()
+        self._ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        self.connect()
+        if self.use_sftp:
+            self._sftp = self._ssh.open_sftp()
```

Comment:
May be better make _sftp property like this:
```
@property
def _sftp(self):
    if self._sftp_client is None:
        self._sftp_client = self._ssh.open_sftp()
    return self._sftp_client
```
In such case it won't initialized until it needed.